traits/src/currency.rs

```diff
@@ -288,6 +288,77 @@ pub trait NamedMultiReservableCurrency<AccountId>: MultiReservableCurrency<Accou
 		value: Self::Balance,
 		status: BalanceStatus,
 	) -> result::Result<Self::Balance, DispatchError>;
+
+	/// Ensure the reserved balance is equal to `value`.
+	///
+	/// This will reserve extra amount of current reserved balance is less than
+	/// `value`. And unreserve if current reserved balance is greater than
+	/// `value`.
+	fn ensure_reserved_named(
+		id: &Self::ReserveIdentifier,
+		currency_id: Self::CurrencyId,
+		who: &AccountId,
+		value: Self::Balance,
+	) -> DispatchResult {
+		let current = Self::reserved_balance_named(id, currency_id, who);
+		match current.cmp(&value) {
+			Ordering::Less => {
+				// we checked value > current
+				Self::reserve_named(id, currency_id, who, value - current)
+			}
+			Ordering::Equal => Ok(()),
+			Ordering::Greater => {
+				// we always have enough balance to unreserve here
+				Self::unreserve_named(id, currency_id, who, current - value);
+				Ok(())
+			}
+		}
+	}
+
+	/// Unreserve all the named reserved balances, returning unreserved amount.
+	///
+	/// Is a no-op if the value to be unreserved is zero.
+	fn unreserve_all_named(
+		id: &Self::ReserveIdentifier,
+		currency_id: Self::CurrencyId,
+		who: &AccountId,
+	) -> Self::Balance {
+		let value = Self::reserved_balance_named(id, currency_id, who);
+		Self::unreserve_named(id, currency_id, who, value);
+		value
+	}
+
+	/// Slash all the reserved balance, returning the negative imbalance
+	/// created.
```

Comment:
current implementation of `slash_reserved_named` only return one value, not returning nagative imbalance.
comment here can change to: returning the amount that was unable to be slashed.

or we could align at substrate: https://github.com/paritytech/substrate/blob/master/frame/support/src/traits/tokens/currency/reservable.rs#L124 which return tuple in `slash_reserved_named`, but I think this can be a later issue.